hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/util/TokenRelListCalculator.java

```diff
@@ -0,0 +1,323 @@
+        // return value
+        final TokenRelation newNextTokenRel = nextTokenRel != null
+                ? nextTokenRel.copyBuilder().previousToken(prevTokenRelTokenNum).build()
+                : null;
+
+        return new TokenRelPointerUpdateResult(newPrevTokenRel, newNextTokenRel);
+    }
+
+    /**
+     * Given an account's current head token number and a collection of token relations to remove,
+     * this method computes the expected new head token number for the account
+     *
+     * @param currentHeadTokenNum the account's current head token number, i.e. the head token number that may change
+     * @param accountId the account's ID
+     * @param tokenRelsToDeleteByTokenId a map of token relations to delete, keyed by token ID for convenience of lookup
+     * @return the new head token number for the account
+     */
+    private long calculateHeadTokenAfterDeletions(
+            final long currentHeadTokenNum,
+            final AccountID accountId,
+            final Map<Long, TokenRelation> updatedTokenRels,
+            final Map<Long, TokenRelation> tokenRelsToDeleteByTokenId) {
+        // Calculate the new head token number by walking the linked token rels until we find a token rel that is not in
+        // the list of token rels to delete
+
+        var currentTokenNum = currentHeadTokenNum;
+        TokenRelation currentWalkedTokenRel;
+        do {
+            currentWalkedTokenRel = updatedTokenRels.containsKey(currentTokenNum)
+                    ? updatedTokenRels.get(currentTokenNum)
+                    : tokenRelStore
+                            .get(
+                                    accountId,
+                                    TokenID.newBuilder()
+                                            .tokenNum(currentTokenNum)
+                                            .build())
+                            .orElse(null);
+            if (currentWalkedTokenRel != null) {
+                if (!tokenRelsToDeleteByTokenId.containsKey(currentWalkedTokenRel.tokenNumber())) {
+                    // we found the first existing token rel that is not in the list of token rels to delete
+                    break;
+                } else {
+                    // we found a non-null token rel, but it is in the list of token rels to delete; we therefore
+                    // continue walking the linked token rels
+                    currentTokenNum = currentWalkedTokenRel.nextToken();
+                }
+            } else {
+                // We reached the end of the linked token rel pointers chain; there is no token rel that will qualify as
+                // the new head token number. We therefore set the new head token number to -1 and exit the do-while
+                // loop (since `currentWalkedTokenRel` is null)
+                currentTokenNum = -1;
+            }
+        } while (currentWalkedTokenRel != null);
```

Comment:
Imo it's good to bound all such linked list traversals with a counter, since in theory a bug could cause the `nextToken()` of a rel to point to itself and hence make this an infinite loop. 😰 

We do track the [total # of associations](https://github.com/hashgraph/hedera-protobufs/blob/add-pbj-types-for-state/services/state/token/account.proto#L119) for an account, so that would be a good bound.

Reply:
I didn't think about this, great idea. Power to the paranoid!

Reply:
Done